src/vmm/src/rpc_interface.rs

```diff
@@ -408,6 +407,16 @@ impl<'a> PrebootApiController<'a> {
             }
         }
 
+        // TODO: This eventfd no affect on the pre-boot API, we only read on this eventfd such that
+        // when it is passed to the post-boot API it is zero (the API thread always writes to the
+        // eventfd and does not differentiate between pre-boot and post-boot). Here we are making
+        // unnecessary syscalls that slow down the API. This should be removed.
```

Comment:
`write()` syscall on eventfd is really cheap and there are only a few calls during VM setup - the optimization you speak of here has zero practical impact on VM setup/boot latency - I will be very surprised if you could actually measure any difference that is not simple noise.

On the other hand, if I understand your comment, the cost of making this optimization is making `api_server` aware of which commands are before boot and which are after which is outside its logical scope and would brake the components' current decoupled design.

I agree this is all more complex than it could be, but the reason for it was to have a VMM that can be built without api server (see https://github.com/firecracker-microvm/firecracker/pull/1489 and https://github.com/firecracker-microvm/firecracker/pull/1490).

Now, if you want to tie back the components together and remove the generic rpc layer, I suggest you do just that, and completely remove this `rpc_interface` layer and just move the logic to api_server.

Reply:
Maybe the diagrams in https://github.com/firecracker-microvm/firecracker/pull/1490 description would have also been useful to be included as docs in the codebase.